source/extensions/io_socket/user_space/client_connection_factory.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/network/client_connection_factory.h"
+#include "envoy/network/connection.h"
+
+#include "source/common/common/logger.h"
+#include "source/extensions/io_socket/user_space/thread_local_registry.h"
+
+namespace Envoy {
+
+namespace Extensions {
+namespace IoSocket {
+namespace UserSpace {
+
+class InternalClientConnectionFactory : public Network::ClientConnectionFactory,
+                                        Logger::Loggable<Logger::Id::connection> {
+public:
+  ~InternalClientConnectionFactory() override = default;
+  std::string name() const override { return "EnvoyInternal"; }
+  Network::ClientConnectionPtr
+  createClientConnection(Event::Dispatcher& dispatcher,
+                         Network::Address::InstanceConstSharedPtr address,
+                         Network::Address::InstanceConstSharedPtr source_address,
+                         Network::TransportSocketPtr&& transport_socket,
+                         const Network::ConnectionSocket::OptionsSharedPtr& options) override;
+  // The slot is owned by the internal listener registry extension. Once that extension is
+  // initialized, this slot is available. The ClientConnectionFactory has two potential user cases.
+  // 1. The per worker thread connection handler populates the per worker listener registry.
+  // 2. A envoy thread local cluster lookup the per thread internal listener by listener name.
+  // Since the population and the lookup is supposed to be executed in the same worker thread,
+  // neither need to hold a lock.
+  // TODO(lambdai): make it friend to only bootstrap extension.
+  static ThreadLocal::TypedSlot<Extensions::InternalListener::ThreadLocalRegistryImpl>*
```

Comment:
Why do you need this? Can't you just get the slot from the InternalListenerExtension singleton?

Reply:
Somehow my comment was lost...

This factory is constructed providing very few contexts: the dispatcher as the context is vetoed.
 
The major blocker is that Singleton object cannot be get from worker thread even though we concede that performance won't be an issue.

I am open to any alternative approach!